### PR TITLE
Add plugins support

### PR DESCRIPTION
**Pull request by shlevy**
Alternative solution to #1841 / #1844.

### CODE REVIEW

File: src/build-remote/build-remote.cc [hunk 1]
```diff
@@ -64,6 +65,8 @@ int main (int argc, char * * argv)
 
         settings.maxBuildJobs.set("1"); // hack to make tests with local?root= work
 
+        initPlugins();
```
Comment (edolstra) [inline src/build-remote/build-remote.cc]: Why are all these `initPlugins()` calls necessary? Can't that be done in `initNix()` or some other common location?

Comment (shlevy) [reply src/build-remote/build-remote.cc]: They need to be called after `--option` flags are parsed, which is done in a bunch of different ways by different programs.

File: src/libexpr/primops/fetchMercurial.cc [hunk 2]
```diff
@@ -204,6 +203,11 @@ static void prim_fetchMercurial(EvalState & state, const Pos & pos, Value * * ar
         state.allowedPaths->insert(hgInfo.storePath);
 }
 
-static RegisterPrimOp r("fetchMercurial", 1, prim_fetchMercurial);
+static void reg(EvalState & state)
+{
+    state.addPrimOp("fetchMercurial", 1, prim_fetchMercurial);
+}
+
+static RegisterPlugin r(Plugin(PluginTag<PluginType::EvalPlugin>(), reg));
```
Comment (edolstra) [inline src/libexpr/primops/fetchMercurial.cc]: I don't understand why these changes are necessary. Doesn't the `RegisterPrimOp` constructor run automatically when the plugin is loaded? Why the need for a `RegisterPlugin` with a lot of template crap?

Comment (shlevy) [reply src/libexpr/primops/fetchMercurial.cc]: It's not necessary, this is all contained in the second commit that I explicitly said could be dropped ;) This was just to drop duplicate capabilities.

Also, a single template is hardly "a lot of template crap".

Comment (shlevy) [reply src/libexpr/primops/fetchMercurial.cc]: ... Though I just realized an easy way to avoid the template altogether, patch incoming.

**Comment by shlevy** [comment]
@edolstra Template crap deleted, second commit simplified (though still not needed)

File: src/libexpr/primops/fetchMercurial.cc (hunk 2 above)
Comment (edolstra) [reply src/libexpr/primops/fetchMercurial.cc]: Well, I don't understand the need for all this over-engineering (e.g. `Plugin(PluginTag<PluginType::EvalPlugin>()`), plus all that complexity in `plugins.hh`, which for some reason has a forward-decl to `EvalState`...

File: src/libexpr/primops/fetchGit.cc [hunk 3]
```diff
@@ -238,6 +237,11 @@ static void prim_fetchGit(EvalState & state, const Pos & pos, Value * * args, Va
         state.allowedPaths->insert(gitInfo.storePath);
 }
 
-static RegisterPrimOp r("fetchGit", 1, prim_fetchGit);
+static void reg(EvalState & state)
```
Comment (shlevy) [inline src/libexpr/primops/fetchGit.cc]: This could be inlined with a lambda if desired.

File: src/libexpr/primops/fetchMercurial.cc (hunk 2 above)
Comment (shlevy) [reply src/libexpr/primops/fetchMercurial.cc]: I removed the tag stuff, was a holdover from an earlier approach. `plugins.hh` needs to reference `EvalState &` to definte the `EvalPlugin` type, but #including `eval.hh` this low in the header chain could easily lead to accidental references to libnixexpr in libnixstore, if it even worked otherwise.

**Comment by shlevy** [comment]
@edolstra Continuing out of the source  comments: `plugins.hh` is ready to support multiple plugin types for two reasons: 1) You mentioned wanting to move s3binarycachestore to plugins and 2) it's very cheap forward compatibility.

**Comment by edolstra** [comment]
Well, I still don't understand why RegisterPlugin is needed at all (and why plugin.cc needs to know about EvalState). The *only* thing that should be needed for plugin support is to do `dlopen` on the .so files, the `RegisterPrimOp` and `RegisterStoreImplementation` constructors should take care of the registration.

**Comment by shlevy** [comment]
OK, I'll switch tothat approach.

File: src/buildenv/buildenv.cc [hunk 4]
```diff
@@ -135,6 +136,7 @@ int main(int argc, char ** argv)
 {
     return handleExceptions(argv[0], [&]() {
         initNix();
+        initPlugins();
```
Comment (edolstra) [inline src/buildenv/buildenv.cc]: `buildenv` should not need plugins.

File: src/resolve-system-dependencies/resolve-system-dependencies.cc [hunk 5]
```diff
@@ -158,6 +158,7 @@ int main(int argc, char ** argv)
 {
     return handleExceptions(argv[0], [&]() {
         initNix();
+        initPlugins();
```
Comment (edolstra) [inline src/resolve-system-dependencies/resolve-system-dependencies.cc]: This command probably does not need plugins.

**Comment by shlevy** [comment]
@edolstra updated

**Comment by edolstra** [comment]
Merged, thanks!

Feature request: it would be useful to support loading all .so's in a directory, e.g.
```
plugin-files = ${config.system.path}/lib/nix/plugins
```
to load all plugins provided by installed packages.
